### ISSUE TITLE
graph::CreateUser - Unable to specify correct UserCreationContext

### ISSUE THREAD

**Issue by tiwood** [microsoft/azure-devops-go-api#33]
To create a user with `graph.CreateUser` we have to provide `CreateUserArgs`:

```go
// Arguments for the CreateUser function
type CreateUserArgs struct {
	// (required) The subset of the full graph user used to uniquely find the graph subject in an external provider.
	CreationContext *GraphUserCreationContext
	// (optional) A comma separated list of descriptors of groups you want the graph user to join
	GroupDescriptors *[]string
}
```
 
The CreationContext decides if we create an `aad`, `ad` or `msa` account, but the current implementation only allows us to use the following context (which shouldn't be used):

```go
// Do not attempt to use this type to create a new user. Use one of the subclasses instead. This type does not contain sufficient fields to create a new user.
type GraphUserCreationContext struct {
	// Optional: If provided, we will use this identifier for the storage key of the created user
	StorageKey *uuid.UUID `json:"storageKey,omitempty"`
}
```

Instead we should be able to provide one of the following:
```go
type GraphUserPrincipalNameCreationContext struct {
	// Optional: If provided, we will use this identifier for the storage key of the created user
	StorageKey *uuid.UUID `json:"storageKey,omitempty"`
	// This should be the principal name or upn of the user in the source AD or AAD provider. Example: jamal@contoso.com Team Services will communicate with the source provider to fill all other fields on creation.
	PrincipalName *string `json:"principalName,omitempty"`
}

type GraphUserMailAddressCreationContext struct {
	// Optional: If provided, we will use this identifier for the storage key of the created user
	StorageKey  *uuid.UUID `json:"storageKey,omitempty"`
	MailAddress *string    `json:"mailAddress,omitempty"`
}
```

**Comment by tedchamb** [comment]
Thanks @tiwood , this is an issue where Go's lack of inheritance makes this difficult to address in the code generator.  This is going to affect any method that accepts different subclasses as input.  Fortunately there are not a lot of methods that fall into this category.  I have a plan on how to fix this, but not sure when I will be able to get to it.  Can you give me an idea of the priority of this particular issue for your application?

Thanks,
Ted

**Comment by tedchamb** [comment]
The [`GraphClient`](https://github.com/microsoft/azure-devops-go-api/blob/dev/azuredevops/graph/client.go) is meant to replace the [`IdentityClient`](https://github.com/microsoft/azure-devops-go-api/blob/dev/azuredevops/identity/client.go), but the [`IdentityClient`](https://github.com/microsoft/azure-devops-go-api/blob/dev/azuredevops/identity/client.go) still works.  Until this issue is fixed, you may want to try the `CreateIdentity` method on the [`IdentityClient`](https://github.com/microsoft/azure-devops-go-api/blob/dev/azuredevops/identity/client.go).

**Comment by tmeckel** [comment]
@tedchamb  is it possible to change the signature of the CreateUser method as follows (making parameter `args` generic)?

```go
func (client *ClientImpl) CreateUser(ctx context.Context, args interface{}) (*GraphUser, error) {
``` 

and then checking in the method body for valid struct types?

**Comment by tmeckel** [comment]
> 
> 
> The [`GraphClient`](https://github.com/microsoft/azure-devops-go-api/blob/dev/azuredevops/graph/client.go) is meant to replace the [`IdentityClient`](https://github.com/microsoft/azure-devops-go-api/blob/dev/azuredevops/identity/client.go), but the [`IdentityClient`](https://github.com/microsoft/azure-devops-go-api/blob/dev/azuredevops/identity/client.go) still works. Until this issue is fixed, you may want to try the `CreateIdentity` method on the [`IdentityClient`](https://github.com/microsoft/azure-devops-go-api/blob/dev/azuredevops/identity/client.go).

@tedchamb I checked the implementation of the `IdentityClient` at https://github.com/microsoft/azure-devops-go-api/blob/273e55a7119a766faedd7a951bc1894171f89bd9/azuredevops/identity/client.go#L158 but the GUID is no more listed in the API endpoint which could be downloaded via the  `_apis` endpoint. So the `IdentityClient` will yield a `LocationIdNotRegisteredError` from here https://github.com/microsoft/azure-devops-go-api/blob/273e55a7119a766faedd7a951bc1894171f89bd9/azuredevops/client.go#L208

Same with the GUID which might be used for creating groups, 

https://github.com/microsoft/azure-devops-go-api/blob/273e55a7119a766faedd7a951bc1894171f89bd9/azuredevops/identity/client.go#L133

I wanted to use the `IdentityClient` because the graph api implementation for groups has the same bug as described above by @tiwood for users. You simply can't create a new groups because you can't pass the required structs to `CreateGroup` at https://github.com/microsoft/azure-devops-go-api/blob/273e55a7119a766faedd7a951bc1894171f89bd9/azuredevops/graph/client.go#L153 

For reference, this is the Powershell code I used to validate if the required API GUIDs are available via the `/_apis` endpoint

```Powershell
$orgUrl = 'https://dev.azure.com/ophiosdev'
$pat = ''

$argv = @{
    Uri = $orgUrl + '/_apis'
    Method = 'OPTIONS'
    Headers = @{
        "Accept" = "application/json"
        "Authorization" = "Basic " + [convert]::ToBase64String([Text.Encoding]::ASCII.GetBytes($pat))
    }
}
$apiMap = (Invoke-RestMethod @argv).Value `
| ForEach-Object -Begin {
    $_map = @{}
} -Process {
    $_map[$_.id] = $_
} -End {
    $_map
}

$apiMap["dd55f0eb-6ea2-4fe4-9ebe-919e7dd1dfb4"] # User Identity API
$apiMap["5966283b-4196-4d57-9211-1b68f41ec1c2"] # Group Identity API
```

**Comment by tmeckel** [comment]
> 
> 
> The [`GraphClient`](https://github.com/microsoft/azure-devops-go-api/blob/dev/azuredevops/graph/client.go) is meant to replace the [`IdentityClient`](https://github.com/microsoft/azure-devops-go-api/blob/dev/azuredevops/identity/client.go), but the [`IdentityClient`](https://github.com/microsoft/azure-devops-go-api/blob/dev/azuredevops/identity/client.go) still works. Until this issue is fixed, you may want to try the `CreateIdentity` method on the [`IdentityClient`](https://github.com/microsoft/azure-devops-go-api/blob/dev/azuredevops/identity/client.go).

@tedchamb as we have the same issue with creating Groups via the Graph API I wanted to use the `CreateGroups` function from the `Indentity` API.  The following structure must be passed as argument to `CreateGroups`

```go
// Arguments for the CreateGroups function
type CreateGroupsArgs struct {
	// (required)
	Container interface{}
}
```

But because an `interface{}` is used as type for the `Container` struct element and there's no further description on this, I'm unable to use the `CreateGroups` method.

**Comment by nechvatalp** [comment]
I attempted to fix this in the latest 7.1 release by providing by providing more variants of the methods each accepting different context.

I added such "overloads" for both Users and Groups, I was able to call create user successfully with that.
